### DISCUSSION TITLE
Issue copying data from auth.user.raw_user_meta_data to a public table via a function.

### DISCUSSION THREAD

**Discussion by taylorcantwell** [supabase/supabase#4280] [Questions]
Hi, I'm attempting to copy certain signup information to a user table on the public schema. However, I'm stuck with copying data from the raw_user_meta_data column. The method that I'm trying to replicate is from https://github.com/vercel/nextjs-subscription-payments/blob/767eb938e541c635489703a5d8921b6e12aceb20/schema.sql#L26

The trigger and function is partially working in that the email and id is copied over to the user table, but as soon as I try to extract some data from the  raw_user_meta_data column, it returns this following error: "code":500,"msg":"Database error saving new user" and nothing is saved at all.

EDIT
The json (raw_user_meta_data) is being stored in the the auth schema on a successful sign up (successful sign up being when I'm not trying to copy this field to the public table).

```
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();
```

The function:

```
create or replace function public.handle_new_user() 
returns trigger as $$
begin
  insert into public."User" (id, email, username)
  values (new.id, new.email, new.raw_user_meta_data ->>'username');
  return new;
end;
$$ language plpgsql security definer;
```

public.User.username is text, and non nullable.

Any help would be much appreciated and thank in advance.

**Comment by GaryAustin1** [comment]
Are you guaranteed {username:'something'} is in the json?  Not sure what value is returned if not (null? or error).

**Reply by taylorcantwell** [reply]
Yep. The json (raw_user_meta_data) is being stored in the the auth schema on a successful sign up (successful sign up being when I'm not trying to copy this field to the public table).

**Reply by GaryAustin1** [reply]
I'm not expert on Postgres json so not much more help.  If you haven't already, you probably should do some straight sql in the sql window selecting and inserting the data with your json code so you can at least see if there are errors that you can't see with the trigger function.  I'm pretty sure ->> returns text so should not need to type cast, unless you need to do that on new.raw_user_meta_data::jsonb or some such but again I don't think so.

**Comment by bastiaanv** [comment]
I am a bit late to the party, but GoTrue does not use `username` to do authentication, so it does not exists in the `raw_user_meta_data`. You can try to grab the `email` property to fetch the users “username”.

Example:
```
create or replace function public.handle_new_user() 
returns trigger as $$
begin
  insert into public."User" (id, email, username)
  values (new.id, new.email, new.raw_user_meta_data ->>'email');
  return new;
end;
$$ language plpgsql security definer;
```

**Comment by Prantoe** [comment]
```
BEGIN
  INSERT INTO userprofile (id, username, name, role) 
  VALUES 
  (new.id, new.email, new.raw_user_meta_data ->>'name', (new.raw_user_meta_data::json->>'role')::user_role);
    return new;
    END;
```

**Reply by tomasfrancisco** [reply]
Thank you for sharing. I just want to mention that this detail with the `()::cast` made a whole difference for me:

`(` `new.raw_user_meta_data::json->>'role'` `)` `::user_role)`
